src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/markduplicates/MarkDuplicatesFromShardsDataflowTransform.java

```diff
@@ -0,0 +1,165 @@
+        LocalCollection<GATKRead> fragments = readsPartitioned.get(MarkDuplicates.ReadsPartition.PRIMARY.ordinal());
+        LocalCollection<GATKRead> fragmentsTransformed = transformFragments(header, fragments);
+
+        LocalCollection<GATKRead> pairs = readsPartitioned.get(MarkDuplicates.ReadsPartition.PRIMARY.ordinal());
+        LocalCollection<GATKRead> pairsTransformed = transformReads(header, finder, pairs);
+
+        LocalCollection<GATKRead> not_primary = readsPartitioned.get(MarkDuplicates.ReadsPartition.NOT_PRIMARY.ordinal());
+
+        LocalCollection<GATKRead> ret = LocalCollection.union(fragmentsTransformed, pairsTransformed, not_primary)
+            .map(MarkDuplicatesFromShardsDataflowTransform::removeHeader);
+
+        return ret;
+    }
+
+
+    private static LocalCollection<GATKRead> transformFragments(final SAMFileHeader header, LocalCollection<GATKRead> fragments) {
+        return fragments
+            // group by key (genomic location etc)
+            .groupBy((read) -> MarkDuplicatesUtils.keyForFragment(header, read))
+            // in each group, mark the duplicates.
+            // false = do not make a copy. That's safe here because we know exactly how LocalCollection works.
+            .mapTransform((reads)->MarkDuplicatesUtils.transformFragmentsFn(reads,false))
+            // go back to a single list
+            .flatten();
+    }
+
+    /**
+     * (1) keyReadsByName: label each read with its read group and read name.
+     * (2) GroupByKey: group together reads with the same group and name.
+     * (3) keyPairedEndsWithAlignmentInfo:
+     *   (a) Sort each group of reads (see GATKOrder below).
+     *   (b) Pair consecutive reads into PairedEnds. In most cases there will only be two reads
+     *       with the same name. TODO: explain why there might be more.
+     *   (c) Label each read with alignment information: Library, reference index,
+     *       stranded unclipped start and reverse strand.
+     *   (d) Leftover reads are emitted, unmodified, as an unpaired end.
+     * (4) GroupByKey: Group PairedEnds that share alignment information. These pairs
+     *     are duplicates of each other.
+     * (5) markDuplicatePairs:
+     *   (a) For each group created by (4), sort the pairs by score and mark all but the
+     *       highest scoring as duplicates.
+     *   (b) Determine which duplicates are optical duplicates and increase the overall count.
+     */
+    static LocalCollection<GATKRead> transformReads(final SAMFileHeader header,
+                                                    final OpticalDuplicateFinder finder, final LocalCollection<GATKRead> pairs) {
+
+        return pairs
+            // only keep the reads that have a mapped mate
+            .filter(ReadUtils::readHasMappedMate)
+            // group reads by genomic location etc.
+            .groupBy((read) -> MarkDuplicatesUtils.keyForRead(header, read))
```

Comment:
I think you have shard boundary issues here too -- what's to prevent two reads with the same name from being assigned to different shards? The original implementation did not have this issue, since it did a global group-by on the read name.